operator: Document how to connect to distributor component (#5551)
grafana/loki: operator/docs/forwarding_logs_without_gateway.md

    spec:
      collection:
        logs:
          type: "vector"
          fluentd: {}
    ```

* By default, TLS is enabled on all components deployed by loki-operator. Because the service certificates are signed by a cluster-internal CA you need to set up a secret that enables the collector to validate the certificate returned by the distributor. The secret must exist in the openshift-logging namespace, and must have a key `ca-bundle.crt`.

  The CA certificate is part of a ConfigMap that gets created by loki-operator as part of the LokiStack. Unfortunately this ConfigMap can not be used directly and has to be converted to a Secret readable by the collector.

    Fetch the `ca-bundle.crt` using:

    ```console
    kubectl -n openshift-logging get cm lokistack-dev-gateway-ca-bundle -o jsonpath="{.data.service-ca\.crt}"
    ```
  
* Once secret is fetched, create a new secret file:

    ```console
    kubectl -n openshift-logging create secret generic loki-distributor-ca \
    --from-literal=ca-bundle.crt="<CA_BUNDLE>"
    ```
    
    where `<CA_BUNDLE>` is the copied value from previous step.

* Now create a ClusterLogForwarder CR to forward logs to LokiStack:

  ```yaml

>   apiVersion: "logging.openshift.io/v1"
>   kind: "ClusterLogForwarder"
>   metadata:
>     name: "instance"
>     namespace: "openshift-logging"

Reviewer: ```suggestion
  apiVersion: logging.openshift.io/v1
  kind: ClusterLogForwarder
  metadata:
    name: instance
    namespace: openshift-logging
```

>   apiVersion: "logging.openshift.io/v1"
>   kind: "ClusterLogForwarder"
>   metadata:
>     name: "instance"
>     namespace: "openshift-logging"

Reviewer: ```suggestion
  apiVersion: logging.openshift.io/v1
  kind: ClusterLogForwarder
  metadata:
    name: instance
    namespace: openshift-logging
```

  spec:
    outputs:
     - name: loki-operator

>        type: "loki"

Reviewer: ```suggestion
       type: loki
```

       url: https://lokistack-dev-distributor-http.openshift-logging.svc:3100
       secret:
         name: loki-distributor-ca
       loki:
         tenantKey: log_type
    pipelines:
     - name: send-logs
       inputRefs:
       - application
       - audit
       - infrastructure
       outputRefs:
       - loki-operator
  ```

  _Note:_ The `tenantKey: log_type` gets resolved as `application`, `audit` or `infrastructure` by fluentd and vector based on the type of logs being collected. This is later used as different tenants for storing logs in Loki.

> ## Troubleshooting

Reviewer: Should we mention that `tenantKey` is currently broken when using a `fluentd` collector?

Author: I don't think we need to mention such things in doc. Also, that bug might be resolved by the time release happens.

# Forwarding Logs to LokiStack without LokiStack Gateway

This document will describe how to send application, infrastructure, and audit logs to the Loki Distributor as different tenants using Fluentd or Vector.

__Please read the [hacking guide](./hack_loki_operator.md) before proceeding with the following instructions.__

_Note:_ This document is meant to be used with OpenShift only.

_Disclaimer:_ This document helps in connecting fluentd or vector to LokiStack without authentication/authorization.

## Openshift Logging

[Openshift Logging](https://github.com/openshift/cluster-logging-operator) supports [forwarding logs to an external Loki instance](https://docs.openshift.com/container-platform/4.9/logging/cluster-logging-external.html#cluster-logging-collector-log-forward-loki_cluster-logging-external) using fluentd and vector as log forwarders.
The below step-by-step guide will help you to send application, infrastructure, and audit logs to the LokiStack through the Distributor endpoint.
The steps remain same for both fluentd and vector.

In order to enable communication between the client(s) and the gateway, follow these steps:

> * Deploy the Loki Operator and a `lokistack` instance for [OpenShift](./hack_loki_operator.md#hacking-on-loki-operator-on-openshift).

Author: We have followed the current pattern everywhere in all howto. Let's keep that itself for consistency as of now.

* Deploy the OpenShift Logging Operator from the Operator Hub or using the following command locally:

    ```console
    make deploy-image deploy-catalog install
    ```
  

> * Create a Cluster Logging instance in the `openshift-logging` namespace with only `collection` defined.

Reviewer 2: Probably should supply a link for how to deploy Cluster Logging.

Author: Since it is for openshift only so they would know the steps or check the official openshift docs.

    For fluentd:

    ```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: "ClusterLogging"
    metadata:
      name: "instance"
      namespace: openshift-logging
    spec:
      collection:
        logs:
          type: "fluentd"
          fluentd: {}
    ```

    For vector:

    ```yaml
    apiVersion: "logging.openshift.io/v1"
    kind: "ClusterLogging"
    metadata:
      name: "instance"
      namespace: openshift-logging
    spec:
      collection:
        logs:
          type: "vector"
          fluentd: {}
    ```

> * By default, the TLS communication is on for communication with the distributor. Therefore, you need to create a secret. The secret must exist in the openshift-logging namespace, and must have a key `ca-bundle.crt`.

Reviewer: ```suggestion
* By default, TLS is enabled on all components deployed by loki-operator. Because the service certificates are signed by a cluster-internal CA you need to set up a secret that enables the collector to validate the certificate returned by the distributor. The secret must exist in the openshift-logging namespace, and must have a key `ca-bundle.crt`.
```

>     This key already exists when LokiStack instance got created but as a Config Map. Hence, you need to fetch them and put into a secret file.

Reviewer: ```suggestion
    The CA certificate is part of a ConfigMap that gets created by loki-operator as part of the LokiStack. Unfortunately this ConfigMap can not be used directly and has to be converted to a Secret readable by the collector.
```

    Fetch the `ca-bundle.crt` using:

    ```console
    kubectl -n openshift-logging get cm lokistack-gateway-lokistack-dev-ca-bundle -o jsonpath="{.data.service-ca\.crt}" | base64 -w 0
    ```
  
* Once all the secrets are fetched, create a new secret file:

    ```yaml
    apiVersion: v1
    kind: Secret
    metadata:

>       name: loki-distributor-metrics

Reviewer 2: Shouldn't this be the <SECRET_FILE_NAME>

      namespace: openshift-logging
    type: Opaque
    data:
      ca-bundle.crt: <CA_BUNDLE>
    ```
    
    where `<CA_BUNDLE>` is the copied value from previous step.

* Create the above secret file in the cluster:

    ```console

>     kubectl create -f <SECRET_FILE_NAME>

Reviewer: Is it maybe simpler to write the CA to a file and use `oc create secret generic $name --from-file` instead? That way kubectl is also doing the base64-encoding.

    ```

* Now create a ClusterLogForwarder CR to forward logs to LokiStack:

  ```yaml
  apiVersion: "logging.openshift.io/v1"
  kind: "ClusterLogForwarder"
  metadata:
    name: "instance"
    namespace: "openshift-logging"
  spec:
    outputs:
     - name: loki-operator
       type: "loki"
       url: https://loki-distributor-http-lokistack-dev.openshift-logging.svc:3100
       secret:

>          name: loki-distributor-metrics

Reviewer 2: Shouldn't this be the `<SECRET_FILE_NAME>`

       loki:
         tenantKey: log_type
    pipelines:
     - name: send-logs
       inputRefs:
       - application
       - audit
       - infrastructure
       outputRefs:
       - loki-operator
  ```

>   _Note:_ The `tenantKey: log_type` gets resolved as `application`, `audit` or `infrastructure` by fluentd and vector based on the type of logs being collected. This is later used as different tenants for storing logs in Loki.

Reviewer: We should probably mention that while these tenants look identical to the ones that the gateway uses in the OpenShift configuration but do not map to the tenants used by the gateway, because it uses UUIDs internally.

Author: This won't be required once #5624 is merged.

Reviewer: That was great timing :slightly_smiling_face:
